### PR TITLE
CuPy external memory pool with memory pool and function pointers

### PR DESCRIPTION
**Pull request by hvy**
Allow setting external allocators to CuPy via raw pointers. See the docs for its interface. https://github.com/cupy/cupy/pull/1904/files#diff-0cb5dfa3212da5a58c698548f2df2f68R1179 .

### CODE REVIEW

File: cupy/cuda/memory.pyx [hunk 1]
```diff
@@ -1137,3 +1138,73 @@ cdef class MemoryPool(object):
...
+    memory was allocated.
+
+    Args:
+        param_ptr (int): Address of *param*.
+        malloc_ptr (int): Address of *malloc*.
+        free_ptr (int): Address of *free*.
+
+    """
+
+    def __init__(self, param_ptr, malloc_ptr, free_ptr):
```
[the remark this answers is not in the document]
Comment (hvy) [reply cupy/cuda/memory.pyx]: Thanks for the suggestion. Let me change it to `param` as I think `data` can be associated with already existing concepts of ndarrays.

[the remark this answers is not in the document]
Comment (hvy) [reply cupy/cuda/memory.pyx]: I thought it was clear that "malloc" is a function but I can change it to `malloc_func` or maybe better `malloc_func_ptr` as the former is used as the function signature typedef.

[the remark this answers is not in the document]
Comment (hvy) [reply cupy/cuda/memory.pyx]: Or `malloc_func` (typedef) -> `malloc_func_type` and `malloc_ptr` -> `malloc_func`.

[the remark this answers is not in the document]
Comment (hvy) [reply cupy/cuda/memory.pyx]: Fixed.

[the remark this answers is not in the document]
Comment (hvy) [reply cupy/cuda/memory.pyx]: Fixed.

[the remark this answers is not in the document]
Comment (hvy) [reply cupy/cuda/memory.pyx]: I keeping a reference to the `owner` sufficient for that purpose?

**Comment by hvy** [comment]
jenkins, test this please.

File: cupy/cuda/memory.pyx (hunk 1 above)
[the remark this answers is not in the document]
Comment (asi1024) [reply cupy/cuda/memory.pyx]: Added owner argument.

**Comment by asi1024** [comment]
LGTM.

**Pull request merged**